docs/articles/persistence/custom-persistence-provider.md

```diff
@@ -0,0 +1,467 @@
+
+## Implementing Akka.Persistence Journal
+
+Akka.Persistence journal is responsible for storing all events for playback in the event of a recovery.
+
+All of the code examples in this documentation will assume a SQLite database. You can view the complete sample project in `/src/examples/Akka.Persistence.Custom`.
+
+The `Journal` table schemas we will be using are:
+
+[!code-csharp[Schema](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=schema "Journal table schemas")]
+
+### Required Method Implementations
+
+#### ReplayMessagesAsync
+
+```c#
+Task ReplayMessagesAsync(
+    IActorContext context, 
+    string persistenceId, 
+    long fromSequenceNr, 
+    long toSequenceNr, 
+    long max, 
+    Action<IPersistentRepresentation> recoveryCallback)
+```
+
+* **`context`**: The contextual information about the actor processing the replayed messages.
+* **`persistenceId`**: Persistent actor identifier
+* **`fromSequenceNr`**: Inclusive sequence number where replay should start
+* **`toSequenceNr`**: Inclusive sequence number where replay should end
+* **`max`**: Maximum number of messages to be replayed
+* **`recoveryCallback`**: Called to replay a message, may be called from any thread.
+
+This call is __NOT__ protected with a circuit-breaker.
+
+This method should asynchronously replay persistent messages:
+
+* Implementations replay a message by calling `recoveryCallback`.
+* The returned `Task` must be completed when all messages (matching the sequence number bounds) have been replayed.
+* The `Task` must be completed with a failure if any of the persistent messages could not be replayed.
+
+The `toSequenceNr` is the lowest of what was returned by `ReadHighestSequenceNrAsync` and what the user specified as the recovery `Recovery` parameter. This does imply that this call is always preceded by reading the highest sequence number for the given `persistenceId`.
+
+This call is NOT protected with a circuit-breaker because it may take a long time to replay all events. The plugin implementation itself must protect against an unresponsive backend store and make sure that the returned `Task` is completed with success or failure within reasonable time. It is not allowed to ignore completing the `Task`.
+
+##### Code Sample
+
+`ByPersistenceIdSql` in this example refers to this SQL query statement:
+
+[!code-csharp[ByPersistenceIdSql](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=ByPersistenceIdSql "ByPersistenceIdSql SQL Statement")]
+
+SQLite code example:
+
+[!code-csharp[ReplayMessagesAsync](../../../src/examples/Akka.Persistence.Custom/Journal/SqliteJournal.cs?name=ReplayMessagesAsync "ReplayMessagesAsync method implementation")]
+
+#### ReadHighestSequenceNrAsync
+
+```c#
```

Comment:
Do we leave this as it is or make use of the range selection though the line number may change in the future?

Reply:
see above